internal/datastore/mysql/datastore.go

```diff
@@ -196,7 +196,107 @@ func NewMySQLDatastore(uri string, options ...Option) (*Datastore, error) {
 	return store, nil
 }
 
-func newMySQLExecutor(db *sql.DB) common.ExecuteQueryFunc {
+// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+func (mds *Datastore) SnapshotReader(rev datastore.Revision) datastore.Reader {
+	createTxFunc := func(ctx context.Context) (*sql.Tx, txCleanupFunc, error) {
+		tx, err := mds.db.BeginTx(ctx, mds.readTxOptions)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return tx, tx.Rollback, nil
+	}
+
+	querySplitter := common.TupleQuerySplitter{
+		Executor:         newMySQLExecutor(mds.db),
+		UsersetBatchSize: mds.usersetBatchSize,
+	}
+
+	return &mysqlReader{
+		mds.QueryBuilder,
+		createTxFunc,
+		querySplitter,
+		buildLivingObjectFilterForRevision(rev),
+	}
+}
+
+func noCleanup() error { return nil }
+
+// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+// ReadWriteTx tarts a read/write transaction, which will be committed if no error is
+// returned and rolled back if an error is returned.
+func (mds *Datastore) ReadWriteTx(
+	ctx context.Context,
+	fn datastore.TxUserFunc,
+) (datastore.Revision, error) {
+	var err error
+	for i := uint8(0); i <= mds.maxRetries; i++ {
+		var newTxnID uint64
+		if err = BeginTxFunc(ctx, mds.db, &sql.TxOptions{Isolation: sql.LevelSerializable}, func(tx *sql.Tx) error {
+			_, err := tx.ExecContext(ctx, "SET innodb_lock_wait_timeout=1")
```

Comment:
Does this set the timeout on the lock to 1s? If so, please document that and link to the docs for it

Reply:
Yes it does. The default is 50s, which means you could have interleaved write transactions just sitting there for 50s waiting for what amounts to a deadlock. @vroldanbet is it appropriate to override this setting at the session/transaction layer? Can we even override it per transaction?

Reply:
I've refactored this to move it into an option, PTAL.

Reply:
@jakedt I fail to see practical reasons to wait 50s for a row-lock, so this would seem sensible to me. The [docs](https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_lock_wait_timeout) indicate this can be set at both global and session level. If what you intend is to set it only for that transaction, I'm not sure that's exactly what will happen. The session lasts for as long as the connection is around, which is pooled, and governed by the max lifetime established. So I suspect that state will be set even after that transaction completes, can you check that?

Even if that state _leaked_ outside of the transaction, I think it would be _fine_, but it may make things difficult to troubleshoot in the future if the state only emerges when a specific operation takes place in the backend.

Reply:
I pushed an update this morning which centralizes it on the connection factory. Now all of the sessions will get it applied as soon as the connection opens. The tests verify that it has the intended effect.

Reply:
That's great and should address the problem 👍🏻 But at that point it may be worth recommending it as a global. It could be done perhaps in a migration, or through documentation? I guess since a part of me is concerned about doing that in the factory because it leads to _more time waiting for a connection to be opened and more overhead_, in light of all the contention we are observing on getting connections from the pool.